pandas/_libs/include/pandas/datetime/date_conversions.h

```diff
@@ -18,7 +18,10 @@ int scaleNanosecToUnit(npy_int64 *value, NPY_DATETIMEUNIT unit);
 // up to precision `base` e.g. base="s" yields 2020-01-03T00:00:00Z
 // while base="ns" yields "2020-01-01T00:00:00.000000000Z"
 // len is mutated to save the length of the returned string
-char *int64ToIso(int64_t value, NPY_DATETIMEUNIT base, size_t *len);
+char *int64ToIso(int64_t value,
+                 NPY_DATETIMEUNIT valueUnit,
```

Comment:
I don't think we need to add a new argument here - shouldn't we just be using the `base` argument?

Reply:
Base is the output unit (corresponds to ``date_unit``).

This new parameter is to indicate the input unit, aka the unit of value, (for the ``pandas_datetime_to_datetimestruct`` call).

Previously, this was hard-coded as nano.
https://github.com/pandas-dev/pandas/blob/main/pandas/_libs/src/datetime/date_conversions.c#L48

Reply:
Ah gotcha. In that case though I think we need to take a step back and just pass in the datetimestruct along with its metadata; that would match the existing APIs for numpy like `convert_datetimestruct_to_datetime`

Any idea how hard that would be? A lot of this JSON code is spaghetti code so it may take a bit to unwind, but the current way of unwrapping to an int64 value and passing in the metadata separately is kind of confusing

Reply:
I think the current approach is probably OK. 

numpy stores a dt64 array as an int array + the reso metadata too.
(The struct is probably some sort of convenience thing.)

I wish I could pass the entire ndarray to the datetime serialization func, but ujson seems to only operate element by element.

Reply:
Right technically this implementation will work as long as the numpy internals don't change; which is also unlikely. But from a readability perspective it is very confusing - we've essentially decoupled the int and metadata values but are passing them back into this `int64ToIso` function. The function name would suggest an integer along would suffice but it really is two critical components. 

The semantics need to be cleaned up instead of just bolting more on here

Reply:
Here's for instance how numpy handles this internally. Note that `convert_datetime_to_datetimestruct` is defined in their _datetime.h header, but isn't something we've vendored in our np_datetime.h; we should probably just vendor that

Also keep in mind npy_datetime is just typedef'ed from npy_int64 (see npy_common.h), so we shouldn't have to jump through too many hoops to use their methodology

Reply:
I could pull the ``pandas_datetime_to_datetimestruct`` call out of int64ToIso (and rename that to dtstructToIso or smth).

Does that work?

Reply:
I think what we should do is this:

1. Add the descr to the NpyArrayContext struct (type PyArray_Descr)
2. Create a function with the following signature

```c
static char* npy_datetime_to_iso_string(npy_datetime *dt, JSONTypeContext *ti, size_t *_outLen) { ... }
```

The implementation should be pretty close to NumPy's way of converting a primitive npy_datetime value to a char* string. 

If you assign `pc->PyTypeToUTF8 = npy_datetime_to_iso_string` Step 1 from the type context should give you access to the data type of the array via tc->npyarr->dtype so you don't need to add more state management to this at all

The NpyArrayContext predates my contributions to pandas so don't know the full intent, but I think it was meant to mirror numpys `PyArrayObject` type. For maybe timing reasons the `PyArrayObject` type has the `descr` struct member described in point 1 above

Reply:
Does any of this get easier if https://github.com/numpy/numpy/pull/21199 is merged (once our min version catches up i guess)?

Reply:
Sorry for the late reply on this.

I had a look at the approach you mentioned and it doesn't work since the NpyArrContext is cleared out in ``Object_beginTypeContext`` here at the end of the is datetime block.

https://github.com/pandas-dev/pandas/blob/dec49044fe44f2b4dd8b13f87138f0c80b31fbda/pandas/_libs/src/vendored/ujson/python/objToJSON.c#L1472-L1477

So the dtype will still have to be on the TypeContext or the encoder itself for now.
(I will try to move it back under the NpyArrContext and remove that TODO later.)

Reply:
Ah well that is unfortunate...this stuff really needs an overhaul...in another PR on another day